### PR TITLE
Implement add_new_field to 3d solver

### PR DESCRIPTION
**Pull request by tkarna**
Also includes a couple of bugfixes in the 2D model.

### CODE REVIEW

File: thetis/solver.py [hunk 1]
```diff
@@ -593,9 +604,33 @@ def create_fields(self):
...
         self._isfrozen = True
 
+    def add_new_field(self, function, label, name, filename, shortname=None, unit='-', preproc_func=None):
+        """
+        Add a field to :attr:`fields`.
+
+        :arg function: representation of the field as a :class:`Function`
+        :arg label: field label used internally by Thetis, e.g. 'tracer_2d'
+        :arg name: human readable name for the tracer field, e.g. 'Tracer concentration'
+        :arg filename: file name for outputs, e.g. 'Tracer2d'
```
Comment (jwallwork23) [inline thetis/solver.py:614]: Maybe these examples could be names with 3d in them?

File: thetis/solver2d.py [hunk 2]
```diff
@@ -659,7 +659,8 @@ def create_exporters(self):
                                        self.fields,
                                        field_metadata,
                                        export_type='hdf5',
-                                       verbose=self.options.verbose > 0)
+                                       verbose=self.options.verbose > 0,
+                                       preproc_funcs=self._field_preproc_funcs)
```
Comment (jwallwork23) [inline thetis/solver2d.py:663]: Good catch

**Review by jwallwork23 — CHANGES_REQUESTED** [review]
Looks good! Just one minor comment and the failing tests which look easy to fix.

File: thetis/solver.py [hunk 3]
```diff
@@ -288,24 +289,34 @@ def compute_mesh_stats(self):
...
+        print_output(f'3D mesh: {nlayers} layers, {nprisms} prisms')
+        print_output(f'Horizontal element size: {min_h_size:.2f} ... {max_h_size:.2f} m')
+        print_output(f'Vertical element size: {min_v_size:.3f} ... {max_v_size:.3f} m')
+        print_output(f'Number of 2D elevation DOFs: {dofs_elev2d}')
+        print_output(f'Number of 2D velocity DOFs: {dofs_u2d}')
+        print_output(f'Number of 3D h. velocity DOFs: {dofs_u3d}')
+        print_output(f'Number of 3D v. velocity DOFs: {dofs_w3d}')
+        print_output(f'Number of 3D tracer DOFs: {dofs_t3d}')
+        print_output(f'Number of cores: {self.comm.size}')
+        print_output(f'Tracer DOFs per core: ~{dofs_t3d_core:.1f}')
```
[the remark this answers is not in the document]
Comment (tkarna) [reply thetis/solver.py:319]: Yes, I've added similar stats to the 2d solver now

**Review by jwallwork23 — APPROVED** [review]

**Comment by jwallwork23** [comment]
Thanks!

**Pull request merged**